File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
const maxProfit = function(prices) {
  let profit = 0;
  
  let buyDay = 0;
  let sellDay = 1;
  
  const totalDays = prices.length;
  
  while (sellDay <= prices.length) {
    const buyPrice = prices[buyDay];
    const sellPrice = prices[sellDay];
    
    if (sellPrice > buyPrice) {
      profit = Math.max(profit, sellPrice - buyPrice)
    } else {
      buyDay = sellDay;
    }
    sellDay ++;
  } 
  
  return profit
};

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    max_p = 0
    buy = 0
    sell = 1
    
    while sell < prices.length
        if prices[buy] >= prices[sell]
            buy = sell
        else
            max_p = [max_p, prices[sell] - prices[buy]].max
        end
        sell += 1
    end
    
    max_p
end